gateway/blocks_gateway.go

```diff
@@ -0,0 +1,484 @@
+	}
+
+	return md, f, nil
+}
+
+func (api *BlocksGateway) GetAll(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	n, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return md, nil, err
+	}
+	return md, n, nil
+}
+
+func (api *BlocksGateway) GetBlock(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.File, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	return md, files.NewBytesFile(nd.RawData()), nil
+}
+
+func (api *BlocksGateway) Head(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	lastRoot := lastSeg.Cid()
+
+	nd, err := api.dagService.Get(ctx, lastRoot)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	rootCodec := nd.Cid().Prefix().GetCodec()
+	if rootCodec != uint64(mc.DagPb) {
+		return md, files.NewBytesFile(nd.RawData()), nil
+	}
+
+	// TODO: We're not handling non-UnixFS dag-pb. There's a bit of a discrepancy between what we want from a HEAD request and a Resolve request here and we're using this for both
+	fileNode, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	return md, fileNode, nil
+}
+
+func (api *BlocksGateway) GetCAR(ctx context.Context, path ImmutablePath) (ContentPathMetadata, io.ReadCloser, <-chan error, error) {
+	// Same go-car settings as dag.export command
+	store := dagStore{api: api, ctx: ctx}
+
+	// TODO: When switching to exposing path blocks we'll want to add these as well
+	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	if err != nil {
+		return ContentPathMetadata{}, nil, nil, err
+	}
+
+	md := ContentPathMetadata{
+		PathSegmentRoots: roots,
+		LastSegment:      lastSeg,
+	}
+
+	rootCid := lastSeg.Cid()
+
+	// TODO: support selectors passed as request param: https://github.com/ipfs/kubo/issues/8769
+	// TODO: this is very slow if blocks are remote due to linear traversal. Do we need deterministic traversals here?
```

Comment:
Yes, I remember @Jorropo  poked at this and noted that if we want verifiable client without local blockstores, then we need way more than ordered blocks, we also need to  be able to ask for duplicate blocks (https://github.com/ipfs/specs/pull/330) to avoid keeping everything in memory or on disk.

This is not how anything works today: every CAR client keeps blockstore in some form (even if it is ephemeral in-memory one), and there is no "ordered CAR spec", just IPLD library that does not believe in parallelism. 

If we ever have streaming-without-blockstore figured out, we would either have a new CAR version, or some Gateway IPIP for a  parameter to control response type to force determinism. Then, and only then, it will be safe to adjust `Accept-Ranger` and `Etag`. 

But for now, it is a bag of blocks without any contract. IMO it is better to leverage that and prioritize performance.